README
Device::Blkid::E2fsprogs version 0.42
=====================================

NOTE: This extension only supports the older e2fsprogs versions of libblkid (numbered 1.xx.x)
and not the newer versions now bundled as a part of util-linux-ng (v2.15 or better). In almost
every case you would be advised to use Bastian Friedrich's util-linux-ng based Device::Blkid
module as the newer interface is reportedly backward compatible with the old one.
This module would prove useful in any situation where for any reason you are limited on your
systems to a 1.xx libblkid version which is a part of the e2fsprogs package. With respect to
version numbers, it is the convention that libblkid adopts a version number which matches the
package of which it is a part, so when the library was migrated to the util-linux-ng package
it jumped in version number from 1.41.4 to 2.15. As such, the rule of thumb here is that if
the version of libblkid on your system begins with a '1', it is an e2fsprogs-based version of
the library, whereas those which begin with a '2' are the newer util-linux-ng based builds.
Another key difference is the default location of the cache file between the two packages;
the e2fsprogs-based versions use a default cache file at /etc/blkid/blkid.tab whereas the
util-linux-ng versions of libblkid maintain their default cache file at /etc/blkid.tab.

This version has been implemented somewhat differently than Bastian's util-linux-ng build
of the library. He opted to keep much of his logic and processing in XSUB, mine is done
mostly in C; I have only used XSUB for my straight glue, everything else I kept in C. This is
not to be taken as any opinion of statement on XSUB, its merely a reflection of my own
background and tastes.

This package provides a Perl interface to the e2fsprogs-based versions of libblkid. It does not
support the larger and more robust API which has been added and integrated into the libblkid
library since its inclusion in the util-linux-ng package. That said, the libblkid which now
ships with util-linux-ng is reportedly backward compatible with client code dependant upon the
older, original library which would mean that this module should work with it, albeit with the
more limited selection of API calls. Please see the preceding note for additional details.

Libblkid provides a means of identifying block devices as to their content (such as filesystems)
as well as allowing for the extraction of additional information such as filesystem labels,
volume labels, serial numbers, device numbers, unique identifiers, etc. The libblkid library
maintains a mapping of all of this composite information and maintains its association with
a given block device on the system. Libblkid has become common in modern linux distributions
and can be seen in use in places such as configuration files where device paths had been
formerly used, such as fstab.

The library provides a means of probing block devices for this various meta information as well
as accessing the blkid cache file which stores this data on a local disk. Use of the cache file
can allow an unpriviledged user access to this information even though they might not have
read access to the block device.

I have endeavored to provide a more Perlish interface to the library where possible rather than
simply map the C calls to Perl ones. Most calls return an undef on failure and those which deal
with object allocations will throw a catchable exception when something goes wrong. Perl hash
types are returned as well in a number of cases.

Please read the README file in the package archive for instructions should you encounter any
problems while using this software package.

E2FSPROGS-LIBBLKID VERSION HISTORY

Libblkid first appeared in version 1.33 of the e2fsprogs package back in 2003. The initial
release of the library exposed 17 function calls as part of its API. When libblkid stopped
shipping with the e2fsprogs package, the function count numbered at 25 calls which had been
the case since version 1.40 of the interface. A brief outline of the API history is as
follows:

   v1.33  -  17 calls
   v1.34  -  18 calls
   v1.36  -  21 calls
   v1.38  -  24 calls
   v1.40  -  25 calls

In version 1.36 of libblkid two functions appeared which returned version information about
the library, those were:

   int blkid_parse_version_string(const char *ver_string);
   int blkid_get_library_version(const char **ver_string, const char **date_string);

These calls have been used to provide for dynamic library version detection during the
initial configuration in the package Makefile.PL script. As these calls were not available
prior to version 1.36, dynamic configuration is not yet possible for the two earlier versions
and a default baseline target of v1.33 will be generated.
   
INSTALLATION

This module uses an optional dynamic libblkid version detection mechanism to try and ascertain
which version of libblkid is installed on the target system. It will attempt to detect the
version of the libblkid library installed on your target system and should this fail, it assumes
a pre-1.36 version and builds a default v1.33 baseline which is compatible with all versions
of the library.

To install and begin dynamic version detection:

  perl Makefile.PL
  make
  make test
  make install

===== libblkid version detection =====

I have made use of a customized Devel::CheckLib module and Makefile.PL in an attempt to detect
the version of libblkid currently installed on the target system and to then generate a Perl XS
interface which directly targets and matches the API interface of that libblkid version.

This process is expected to work on all versions of libblkid later than v1.35. In the event that
the target system has a pre-1.36 version of the library, a default baseline target of v1.33 is
configured for build. Should you have any problems with this process, evident either in running
the Makefile.PL or in running make on its resulting Makefile, please see the Makefile.PL as well
as the E2fsprogs.xs file to troubleshoot. If you wish to report any problems with this version
detection, please include any output from the installation process as well as a copy of your
/usr/include/blkid/blkid.h file.
   
DEPENDENCIES

Perl 5.8.0

This module also requires these other modules and libraries:

E2fsprogs v1.33-v1.41

* Note: This package supports both dynamic version detection during the configuration phase
        of running Makefile.PL, as well as manual selection of a build target by version.
        If neither of these are suitable to your needs or there is some sort of problem with
        this on your target system, please refer to the Makefile.PL as well as Devel::CheckLib
        files; both are well documented and relatively easy to modify.

REPORTING A BUG

If you have any problems with the functioning of this library, please contact me at my CPAN
address at mroz@cpan.org. Please be sure and include a list of relevant packages installed on
your system or at least the exact version of e2fsprogs which you have installed, perhaps by
providing the blkid.h file.

Should you find any problems with this package and wish to submit a patch, please
send the unified diff to my CPAN address including full patch instructions as well as a
description of what issue or bug you have found and addressed. Please patch as per recommended
best practices found in the patch(1) man page, under "NOTES FOR PATCH SENDERS".
  
COPYRIGHT AND LICENCE

Copyright (C) 2010, 2011 by Raymond Mroz

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.1 or,
at your option, any later version of Perl 5 you may have available.

THANKS

Thanks to Bastian for all of is work on Device::Blkid, I was especially appreciative of your
POD as libblkid itself is not terribly well documented. While I did make a large number of my
own straight test calls in C against the library your documentation kept me in the right
direction and saved me a load of time. Thanks!

Thanks to David Cantrell, David Golden and Yasuhiro Matsumoto for Devel::CheckLib. I hacked it
up a little bit to manage my dynamic version checks and build, hope you don't mind.

Thanks to Tom Erskine for your insight and experience in all things Perl.

I would also like to thank Larry McInnis for the loan of some hardware on which to develop this
module. Most everything I have had been tied up and developing on the latest and greatest version
of Debian didn't make much sense.

BUGS

No known bugs at this time. That said, this module is largely written in C and does contain
a number of memory allocations. While these allocations are done inside of the libblkid itself,
I do make every attempt to free the memory explicitly when I am done with it. That said, leaks
are possible. Report any issues as is detailed above.

FUTURE

Please note that this module should be considered early release. As such, this information or its
API may change at any time. Please check CPAN for the latest version of this software.


Changes
Revision history for Perl extension Device::Blkid::E2fsprogs.

0.01  Sat Jan  1 11:35:09 AST 2011
	- original version; created by h2xs 1.23 with options
		-A -n Device::Blkid::E2fsprogs

0.10  Tue Jan  4 00:13:24 AST 2011
	- Moving this to version 0.10 after some feverish cleaning
        - Package builds, POD is semi-release worthy
        - Ready for upload on CPAN now as a developer release
        - still requires test script completion

0.12  Tue Jan  4 18:30:46 AST 2011
        - Cleaned up POD typos, finished documenting all functions
        - Added last two function wrappers
                - blkid_get_library_version()
                - blkid_parse_tag_string()
        - Have switched undef return types in object creation functions to
          instead throw exceptions. This happens when the cache, device and
          two iterator objects are requested from the lib.
        - Modified and reogranized README somewhat
        - Removed Data::Dump dependency from Makefile.PL

0.14  Wed Jan  5 19:11:04 AST 2011
        - Additional fixes to POD
        - README updated with build and debug build instructions
        - Makefile.PL, author email corrected
 
0.16  Sun Jan  9 09:16:04 AST 2011
        - Added Devel::CheckLib
        - Modified stock Devel::CheckLib to verify libblkid version < version 2.00.0 (util-linux-ng versions)
        - Some POD and README tweaks
        - Makefile.PL changes for Devel::Checklib (defaults not suitable for my needs)
        - MANIFEST updated with new file additions

0.18  Tue Jan 11 10:29:19 AST 2011
        - Modified Devel::CheckLib to make calls on libblkid to get version data and return a -D compiler
          define used to conditionally compile a version specific target of this package
        - Modified Makefile.PL to work with changes to Devel::CheckLib to support targeted module versions
        - Re-ordered E2fsprogs.xs file. Functions now ordered in order of their appearance by version
        - Added C preprocessor directives to E2fsprogs.xs to provide for targeted version builds
        - Updated README with details about targetting versions and version history of libblkid
        - Updated lib/Device/Blkid/E2fsprogs.pm POD with details about libblkid version history and targeting

0.20  Tue Jan 11 23:23:54 AST 2011
        - Added libblkid version numbers to functions in POD (version which function appeared in libblkid)
        - Cleaned up additional areas of POD in E2fsprogs.pm file.
        - Added two additional cache tests to unit test script
        - Additional cleaning to README
        - Package version bumped to 0.20

0.22  Sat Jan 15 20:25:21 AST 2011
        - Updated Makefile.PL to provide for optional library version detection via user interaction
        - Updated Devel::CheckLib to support new optional library version detection via user interaction
        - Minor updates to POD in E2fsprogs.pm
        - README has been updated with new interactive installation instructions
        - E2fsprogs.xs has had the additional C preprocessor defines added to support versions 1.33-1.35
        - Commented out all but two tests in .t script; these require new approach with dynamic versioning
        - Package version bumped to 0.22

0.24  Sun Jan 16 20:35:09 AST 2011
        - Updated Makefile.PL with META_MERGE param to WriteMakefile()
        - Updated README with additional and more accurate information on installations and creating debug builds
        - Updated POD in E2fsprogs.pm file; blkid is better detailed. Some minor reorganizations as well.
        - Package version bumped to 0.24

0.26  Mon Jan 17 06:45:21 AST 2011
        - Changed Makefile.PL to use its own proper prompt() method rather than my own interactive logic which breaks tests
        - Bumped versions to 0.26

0.28  Wed Jan 19 23:35:21 AST 2011
        - Removed interactive version selection from Makefile.PL
        - Makefile.PL and CheckLib now try detection, on fail build a baseline v1.33 target
        - Updated README to reflect removal of interactive installation
        - Updated POD to remove interactive  installation details
        - Bumped version to 0.28

0.30  Sat Jan 22 18:02:49 AST 2011
        - Changed blkid_get_dev_size wrapper in E2fsprogs.xs so that it now accepts the string name representation of the
          device rather than a POSIX file descriptor to achieve a more Perlish interface
        - The get_dev_size function now returns an exception anytime there is a problem originating from a file descriptor
          allocation.
        - Updated POD details on get_dev_size call changes
        - Added block device test image file to project for unit test scripts ( t/img/ext3.fs )
        - Added 02-dev_size.t test script

0.3001 Sun Jan 23 21:35:12 AST 2011
        - Bugfix/package correction push
        - MANIFEST file correctly updated with new test script
        - New test script, 03-dev_tag.t added
        - Changes file corrected
        - Version bumped to 0.3001

0.32 Mon Jan 24 02:16:53 AST 2011
        - Bumped bugfix release to version 0.33

0.34 Mon Jan 24 23:49:20 AST 2011
        - Cleaned up POD in E2fsprogs.pm, typo corrections, clarifications
        - Added additional tests in 03-dev_tag.t unit test script
        - Version bumped to 0.34

0.36 Tue Jan 25 07:01:54 AST 2011
        - Removed C preprocessor debug code from E2fsprogs.xs
        - Corrected POD typos
        - Removed debug build information from README, POD
        - Removed DEBUG define from Makefile.PL
        - Version bumped to 0.36


Makefile.PL
# Makefile.PL - Device::Blkid::E2fsprogs
#
# Ray Mroz - mroz@cpan.org
# Copyright (C) 2011

use lib qw(inc);
use Devel::CheckLib qw/ :funcs :cflag_defs /
  ;    # See legend below for list of constants imported

use 5.008000;
use ExtUtils::MakeMaker;

# This next section is for Devel::CheckLib. I have hax0r3d this a tad. As well as checking
# for the existance of the blkid library, I am also going to make a call on the lib to get
# version data which is used to set a DEFINE for the C preprocessor. This will provide for
# dynamic version detection of libblkid in target systems. Dynamic version detection is
# optional and can be bypassed during the initial configuration when running this script.

# Version history of e2fsprogs libblkid
#
# Version legend
#
# Versions > 2.xx.x ( > 2000 ) is util-linux-ng (use the Device::Blkid library instead)
# Version 1.40 (140) - 25 API calls
# Version 1.38 (138) - 24 API calls
# Version 1.36 (136) - 21 API calls
# Version 1.34 (134) - 18 API calls - Dynamic version configuration is NOT supported for this version
# Version 1.33 (133) - 17 API calls - Dynamic version configuration is NOT supported for this version

# Note: The blkid_get_library_version() call modifies both its parameters (C strings) and
# returns an 'integerized' version based upon the real, string-based version. We use this
# below to determine and return a version type. All versions above 1.40.1 return a 4 digit
# integer and since the return type is only a single byte in width, we need to drop the
# last digit from versions > 2.00.0. Also note, the get version API calls first appeared
# in version 1.36 of libblkid so dynamic version detection is not supported in versions
# and as such, a baseline version 1.33 compliant target is built which will work with any
# version of the library.

# DEBUGGING and DEBUG builds
# To build a DEBUG build of this module, please refer below to the DEFINE argument to the
# WriteMakefile() function. In that function you will locate two DEFINE arguments, one of
# which is the default argument used in the normal construction of this package, the other
# which is commented out and which has an additional DEBUG related define added to it.
# Refer to the details in that section of the file to produce a DEBUG build if this is
# required.

# Constants imported from Devel::Checklib as 'cflag_defs' representing CFLAG defines which
# are passed as C compiler args when building the module.
# API_DEF_133 => qq/-D__API_1_33/,
# API_DEF_136 => qq/-D__API_1_33 -D__API_1_36/,
# API_DEF_138 => qq/-D__API_1_33 -D__API_1_36 -D__API_1_38/,
# API_DEF_140 => qq/-D__API_1_33 -D__API_1_36 -D__API_1_38 -D__API_1_40/,


my $api_define;

$api_define = check_blkid_version_or_exit(
    function => '
        const char *vers = NULL;
        const char *date = NULL;
        int vernum = 0;
        vernum = blkid_get_library_version(&vers, &date);
        if (vernum >= 2000)
            return(200); /* Childs return value only 1 byte wide */
        if (vernum >= 140)
            return(140);
        if (vernum >= 138)
            return(138);
        if (vernum >= 136)
            return(136); 
        return (-1); /* Default catch, something is borked */
        ',
    incpath => '/usr/include/blkid',
    lib     => 'blkid',
    header  => [ 'blkid.h', 'stdio.h', 'stdlib.h' ],
    blkid   => 1,
);

WriteMakefile(
    NAME         => 'Device::Blkid::E2fsprogs',
    VERSION_FROM => 'lib/Device/Blkid/E2fsprogs.pm',
    PREREQ_PM    => {
        'Exporter'      => '0',
        'Test::More'    => '0',
        'Test::Builder' => '0',
    },
    META_MERGE => {
        resources => {
            repository =>
              'git://github.com/raymroz/Device--Blkid--E2fsprogs.git',
        },
    },
    (
        $] >= 5.005
        ? (
            ABSTRACT_FROM => 'lib/Device/Blkid/E2fsprogs.pm',
            AUTHOR        => 'Raymond Mroz <mroz@cpan.org>'
          )
        : ()
    ),
    LIBS          => ['-lblkid'],
    NEEDS_LINKING => 1,
    DEFINE    => "$api_define",
    INC       => '-I.',
    LICENSE   => 'perl',
    EXE_FILES => [],
    dist      => { COMPRESS => 'gzip -9f', SUFFIX => 'gz' },
    clean => { FILES => 'Device-Blkid-E2fsprogs-*' },
);
